Clean up UX for read and unread posts

andy-armstrong opened the pull request:
### Description
 
[TNL-4557](https://openedx.atlassian.net/browse/TNL-4557)

This is a work-in-progress for cleaning up the UX for showing the read/unread state of discussion posts. The UI is ready for product review and I'll wrap up the tests tomorrow.

### Sandbox

The discussion in the following course has a number of representative posts from a variety of users, so it hopefully demonstrates the full capabilities of this work:

- [x] [https://andy-armstrong.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/discussion/forum/](https://andy-armstrong.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/discussion/forum/)

### Testing
- [x] RTL
- [x] safecommit violation code review process
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @bjacobel 
- [x] Code review: @dianakhuang 
- [x] Doc Review: @catong 
- [x] UX review: @chris-mike 
- [ ] Accessibility review: @clrux 
- [x] Product review: @marcotuts

FYI: @robrap 

### Post-review
- [x] Squash commits

bjacobel commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -417,11 +416,9 @@
                 $('.forum-content').empty().append($templateContent);
-                $('.forum-nav-thread-list a').removeClass('is-active').find('.sr')
-                    .remove();
+                $('.forum-nav-thread-list a').removeClass('is-active').find('.sr').remove();
```
I don't think ESLint is going to like this - more than two chained calls and they have to be split onto separate lines:

```
$('.forum-nav-thread-list a')
  .removeClass('is-active')
  .find('.sr')
  .remove();
```

bjacobel replied:
It might be more than three, though, I forget.

andy-armstrong replied:
Thanks for spotting that. I didn't intend to change that, but did so while cleaning up a merge conflict. I'll revert it.

andy-armstrong commented:
@marcotuts @chris-mike let me know how you would like me to move forward with this PR. I'm happy to make any changes you want, as long as we can get accessibility sign off for them (e.g. for using a tiny font for the labels). Another option would be to let this merge to the feature branch and then continue this discussion as part of landing the feature.

chris-mike commented:
Here's a bulleted list for you.

1. Post I'm currently reading shows a grey. This is not accessible.  Change it to "base" which is the next step darker.

2. Regarding the buttons that read "pinned" "following" and "by: staff".   Can you please remove the box around each (so they do not look like buttons). Remove the bold from the text (font size is fine).  And play with the spacing between the FA icon and the text.  The non buttons should align with the post title.  This applies to instances in both the list of all posts as well in the body of the post for "pinned".

3. I'll defer to @catong on this, but i think the "By: Staff" is unnecessary text. I'd prefer to see just "Staff"

4. There is something wrong with the left-alignment of really long post titles.  They are about two (maybe 3) pixels too far left than the other post titles.  This makes the FA discussion/question icon see too close.  See image.
![screen shot 2016-08-11 at 9 28 09 am](https://cloud.githubusercontent.com/assets/16578372/17590089/3ef34f4a-5fa6-11e6-865a-85e81726a7c7.png)

5. Just a mention that when I created an account for your sandbox yesterday that not everything was unread (which is what i would have expected).  Not sure what this is.

6. I think that the magenta "pinned" is not the right color.  I could be wrong.  It SHOULD be this: Sass: palette(secondary, dark) RGB: rgb(152, 44, 98) HEX: 982c62
Assuming it only shows on white or x-back (brand gray), then it will pass wcag aa.

andy-armstrong commented:
Thanks @chris-mike. I'll make those changes today.

FYI, one state I forgot to include on my sandbox is the Community TA. I've added one now, and you can see how this looks here:

![image](https://cloud.githubusercontent.com/assets/5985072/17591924/3bb25e68-5fae-11e6-82c0-7e4164d2fcf4.png)

Let me know if that color is accessible. The closest pattern library color I can use is ``palette(success, text)`` so I'm switching to that now, but maybe green isn't the best color to use since it usually means success and/or payment.

catong commented:
@andy-armstrong @chris-mike I don't feel strongly about the "By" for the Staff/Community TA indicator, am fine with dropping it. For screen readers, would it still be better to include the word "By" so that "By Staff" or "By Community TA" is what is read?

chris-mike commented:
@catong The SR text can be different from the button. So this is purely a visual evaluation.

chris-mike commented:
@andy-armstrong Apply what I said about "by:staff", "following" and "pinned" to "community TA" as well.  The green color for "community TA" needs to change to (if not already):
Sass: palette(success, text) RGB: rgb(0, 129, 0) HEX: 008100

this color should work on white and x-back (greyscale).

catong commented:
@chris-mike - understood. I guess my request was more for @andy-armstrong :-)

andy-armstrong commented:
@chris-mike @marcotuts @catong I've addressed all of Christine's feedback and updated my sandbox here:

https://andy-armstrong.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/discussion/forum/i4x-edx-eiorguegnru-course-foobarbaz/threads/57ac795b27911a59a2000008

Let me know what you think, and if there's more that needs tweaking.

andy-armstrong commented:
Here's the current state of the UI:

![image](https://cloud.githubusercontent.com/assets/5985072/17602851/7a5721fc-5fdc-11e6-8ef3-a48f71425d10.png)

andy-armstrong commented:
@catong points out that the check box icon for resolved questions used to be shown in green. Currently with my changes, the icon follows all of the others, where it is blue for unread and grey if read. @chris-mike and @marcotuts, let me know if you were expecting this to be green in one or both scenarios.

andy-armstrong commented:
@chris-mike I verified your bug where a new user has both read and unread threads when visiting discussions for the first time:

![image](https://cloud.githubusercontent.com/assets/5985072/17603140/d4e8b490-5fdd-11e6-8207-2d91f09720d1.png)

From the UI, it appears that the problem is that every post with comments has an off-by-one error where the 'X new' count is always one less than the total comment count. I wonder if the number shown for the thread includes the post itself, while the unread comment count only counts the comments. Maybe this is my bug that I'm not interpreting these values correctly.

FYI @robrap @dianakhuang

robrap commented:
FYI: There is a story to drop the post from the overall count: https://openedx.atlassian.net/browse/MA-2559

chris-mike commented:
Sorry... still getting used to our new colors. The "currently reading" is just too dark. It's distracting and sends the wrong message.  Can we make background this color instead:
Sass: palette(primary, base) RGB: rgb(0, 117, 180) HEX: 0075b4
Which means that text needs to be white. See screen.
![screen shot 2016-08-11 at 4 15 05 pm](https://cloud.githubusercontent.com/assets/16578372/17603428/027e61ce-5fdf-11e6-8822-465c1f04b613.png)

Also related to the "currently reading" state, there are 2 additional issues.

1. the "by most votes" count is still black. It needs to be white (also captured in screen shot).
2. If you click on a post, and the post page goes from long to short, then a user will lose sight of the post title they are "currently reading". (see second screen - this falls below the list of posts column).  Therefore, I suggest that moving the post you are "currently reading" to the top of the list (or even to the middle), so that it is always visible.  Goal: make the currently reading post always visible.

![screen shot 2016-08-11 at 4 19 40 pm](https://cloud.githubusercontent.com/assets/16578372/17603538/a32de9b4-5fdf-11e6-91ea-a842cef2b35b.png)

andy-armstrong commented:
@chris-mike @marcotuts @catong I believe that I've addressed all the remaining bugs now. I fixed the following:
 - the issue with incorrectly showing some of a new user's posts as read
 - increased the padding around the icons
 - changed the active post to use a blue background (not black)
 - fixed the color of votes when shown for an active post
 - added a transparent left border for read items, to keep the spacing consistent
 - the overlapping "Add a Post" button in inline discussions
 - the overly squashed labels on posts

I think that's it. Please review again on my sandbox.

Note: I'm not going to mess with trying to make the active item always appear on the screen in this PR. I agree that it is worthwhile but the behavior hasn't changed and I don't want to hold up landing the feature branch.

andy-armstrong commented:
@bjacobel @dianakhuang I've updated the Jasmine tests, so this PR is ready for code review now. I have updated the sandbox here:

https://andy-armstrong.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/discussion/forum/

marcotuts commented:
Awesome! Thanks @andy-armstrong for all these changes. 

I found a few small nits / issues I've outlined below: 

1.) (.post-label related) Report action styling / label on responses seems to not be inheriting the styles appropriately. 
![image](https://cloud.githubusercontent.com/assets/2023680/17607737/addeddb2-5ff5-11e6-8546-10058cfaff51.png)

2.) The course navigation tabs are rendering using different colors when viewed from the discussions tab. (Not sure if this is specific to this PR or the feature branch at large though.) 
Sandbox, viewing Discussions Tab
![image](https://cloud.githubusercontent.com/assets/2023680/17607803/140ca740-5ff6-11e6-82d3-a32f345a9c9f.png)

Sandbox, viewing Course Tab
![image](https://cloud.githubusercontent.com/assets/2023680/17607814/28fd84da-5ff6-11e6-80ec-1e69153c4d88.png)

chris-mike commented:
A small nit from me is the spacing between the post title and the votes count. See screen shot - it looks a little tight when there's a long post title and no pinned/staff/following.  Though I'm not exactly sure these are correlated.

![screen shot 2016-08-12 at 9 03 51 am](https://cloud.githubusercontent.com/assets/16578372/17622981/baf19f28-606b-11e6-892b-ea8ac95dc46c.png)

bjacobel commented on common/static/common/js/discussion/views/discussion_thread_list_view.js:
```diff
@@ -370,20 +376,16 @@
                     );
-                }
-                return content;
+                return $(this.threadListItemTemplate(context).toString());
```
Haven't seen this pattern before - what happens / what's the utility of wrapping a `String` in a call to `$()`?

andy-armstrong replied:
This is the pattern for creating an unattached JQuery element, which you can then append to another view. I don't think it reads very well, but I don't have a better way to handle it. Note that this pattern was already used here, and I just collapsed the code down by moving some of the update logic into the Underscore template.

bjacobel commented on common/static/common/templates/discussion/thread-list-item.underscore:
```diff
@@ -72,7 +72,18 @@
+          <%-
+            StringUtils.interpolate(
+              gettext('{unread_comments_count} new'),
+              {unread_comments_count: unread_comments_count}
```
Nit: the JS styleguide says to infix space in single line objects:

```
{ unread_comments_count: unread_comments_count }
```

but we don't (can't?) lint `.underscore templates` - it would still be good to be consistent.

andy-armstrong replied:
That's how I used to write my code, but our edx lint rules override the object-curly-spacing rule to say that you can't have spaces. I've recently been teaching myself to do it without the spaces!

https://www.npmjs.com/package/eslint-config-edx

I don't remember why we decided to override the AirBnB rule which requires spaces like you say.

bjacobel replied:
Ack, no spaces is what's in the styleguide? I don't remember us discussing it at all, which means I picked that override on my own.... but I prefer spaces??!

Sounds like something we should revisit in the future. 👍 here though.

bjacobel commented on lms/static/sass/discussion/elements/_navigation.scss:
```diff
@@ -42,24 +42,24 @@
-    border-radius: 0;
-    background: transparent;
-    box-shadow: none;
+  box-sizing: border-box;
```
Looks like you indented this from 4 to 2 - would rather see four ([as per styleguide](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#syntax--formatting)) but understand the change if it's for internal consistency - looks like it's been historically mixed in this file.

andy-armstrong replied:
I was making it consistent with the rest of the file. It looks like the only parts that used four spaces are the new lines you added recently. Since I prefer four spaces too, I'll re-indent the whole file to use four.

bjacobel commented on lms/static/sass/discussion/elements/_navigation.scss:
```diff
@@ -288,34 +291,24 @@
   @extend %forum-nav-thread-wrapper;
-  width: 80%;
+  @include margin-left($baseline/2);
+  flex-grow: 1;  // This column should consume all the available space
```
How did the flexbox testing in IE11 go?

andy-armstrong replied:
Errr, I didn't try it yet. Thanks for the reminder!

bjacobel commented:
Looks good to me, a few questions and style nits

andy-armstrong commented:
@chris-mike @marcotuts @catong @bjacobel @dianakhuang I've addressed the last few pieces of feedback and I've squashed my commits. I've also updated my sandbox if you want to see where things stand. Please re-review so I can merge this once I've done testing on IE11.

BTW @marcotuts, I'm not going to fix the Pattern Library issues with the course tabs as that wasn't caused in this branch. I'll make a small PR to fix it on master.

catong commented:
@andy-armstrong The latest updates look great to me!

I think keeping the checkmark for "Mark as Answer" be only blue or grey in keeping with the other icons is good; my only question now is whether we'll be applying the same styling to discussions in mobile?

bjacobel commented:
:shipit:

dianakhuang commented:
Code-wise this looks good 👍

chris-mike commented:
@lizcohen will have thoughts on mobile UI.

chris-mike commented:
Looks great. There is still some weirdness with read/unread. Some things seem not right (e.g., post I created is showing as unread. unless logic has been changed, this should be read for me, though with new posts).

marcotuts commented:
Thanks @andy-armstrong! 👍

andy-armstrong commented:
Good news. The styling all looks correct on IE 11 (although there's something off with the Explore Courses button). I'm going to merge this now.

![image](https://cloud.githubusercontent.com/assets/5985072/17634274/d66dab5a-609f-11e6-8527-6ac15687c85c.png)

andy-armstrong commented:
jenkins run lettuce

andy-armstrong commented:
jenkins run lettuce

robrap commented:
@chris-mike @marcotuts @andy-armstrong: FYIL There is currently a defect where your own comments are not automatically marked as read.  The mobile team in Lahore will be working on this (I think): https://openedx.atlassian.net/browse/TNL-4995

andy-armstrong commented:
jenkins run lettuce
